sdks/python/apache_beam/runners/direct/transform_evaluator.py

```diff
@@ -619,6 +619,7 @@ def start_bundle(self):
         step_name=self._applied_ptransform.full_label,
         state=DoFnState(self._counter_factory),
         user_state_context=self.user_state_context)
+    self.runner.setup()
```

Comment:
This should be called once per instance, not in `start_bundle`.

Reply:
In the DirectRunner, we currently have one instance per bundle.  If we want these instances to be long-living, we need a larger refactoring here.

Reply:
Yes, unfortunately the DirectRunner does not cache instances between bundles. Fortunately this code path is only used by DirectRunner and not the other runners.

Reply:
If `self._perform_dofn_pickle_test` is false, it looks like it is not cloned but comes directly from `self._appliedptransform.transform.dofn`. Is the whole appliedptransform clone per bundle?

Reply:
I added some pdb breakpoints and confirmed that the DoFn is cloned during the test and that the lifecycle methods are called on the clones. See this [thread](https://github.com/apache/beam/pull/7994#discussion_r277110671) for where the cloning occurs.